controllers/etcdcluster_controller.go

```diff
@@ -86,51 +115,131 @@ func (r *EtcdClusterReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error)
 	err := cluster.ValidateCreate()
 	if err != nil {
 		log.Error(err, "invalid EtcdCluster")
-		return ctrl.Result{}, nil
+		return ctrl.Result{}, nil, nil
 	}
 
 	service := &v1.Service{}
-	if err := r.Get(ctx, req.NamespacedName, service); err != nil {
+	if err := r.Get(ctx, name, service); err != nil {
 		if !apierrors.IsNotFound(err) {
 			log.Error(err, "unable to fetch EtcdCluster service")
-			return ctrl.Result{}, err
+			return ctrl.Result{}, nil, err
 		}
 		service = headlessServiceForCluster(cluster)
 		if err := r.Create(ctx, service); err != nil {
 			log.Error(err, "unable to create Service", "service", service.Name)
-			return ctrl.Result{}, err
+			return ctrl.Result{}, nil, err
 		}
 		log.V(1).Info("Created Service", "service", service.Name)
-		return ctrl.Result{}, nil
+		return ctrl.Result{}, &event.ServiceCreatedEvent{Object: cluster, ServiceName: service.Name}, nil
 	}
 	log.V(2).Info("Service exists", "service", service.Name)
 
-	peers := &etcdv1alpha1.EtcdPeerList{}
-	if err := r.List(ctx, peers, client.MatchingFields{clusterNameSpecField: cluster.Name}); err != nil {
-		log.Error(err, "unable to list peers")
-		return ctrl.Result{}, err
-	}
-
-	// Assume we can't communicate with the cluster
-	log.V(1).Info("Unable to communicate with etcd cluster")
-	if cluster.Spec.Replicas != nil && int32(len(peers.Items)) < *cluster.Spec.Replicas {
-
-		// Create more peers
-		peerName := nextAvailablePeerName(cluster, peers.Items)
-		log.V(1).Info("Insufficient peers for replicas, adding new peer",
-			"current-peers", len(peers.Items),
-			"desired-peers", cluster.Spec.Replicas,
-			"peer", peerName)
-		peer := peerForCluster(cluster, peerName)
-		if err := r.Create(ctx, peer); err != nil {
-			log.Error(err, "Failed to create peer",
+	if members == nil {
```

Comment:
why is this condition needed? does this break peer replacement? if we delete a peer - it wouldn't be recreated now right?

Reply:
That's true right now. The idea is that the existing 'scale up' code is only for use in the case where we can't talk to the cluster, and may need to add additional peers to achieve quaracy. (such as in bootstrap or some failure conditions).

I think it's a next step to implement reconciliation when we *can* reach the member list.

Reply:
Maybe add a comment explaining this. Or better still some documentation explaining the bootstrap path vs the established cluster path, because I'm not sure I understand it right now.

Reply:
Sure, let me try to explain here and clarify wording. Then I can roll it into a comment in the code.

> The `EtcdCluster` reconcile function has two modes of operation.
> 
> If it *can* contact the etcd API, it will reconcile the `EtcdPeer` resources in Kubernetes to match the member list in etcd. If those match, it will then consider adding or removing elements to the membership API to bring the size of the cluster in-line with the `EtcdCluster` resource's `spec.replicas` value.
> If it *cannot* contact the etcd API, it will inspect the number of `EtcdPeer` resources in the cluster and if there fewer than the expected number from the `EtcdCluster` resource's `spec.replicas` value, then it will add new peer resources.
> It's important to note that the second mode of operation can never *remove* peer resources. It's only concern is that if etcd is unreachable then it may need to add more peers to make it quorate. This will occur during bootstrap, but may occur during certain failure conditions (typically loss of a peer resource through other means, e.g., an accidental `kubectl etcdpeer delete --all` by an administrator).
> The loop for the first mode of operation will generally occur during scaling. For example if an administrator edited the cluster resource and set the `spec.replicas` of a working cluster from three to five then the following would occur.
> 1. Operator can contact etcd; members match peers; scale too big by two => Add one member to etcd peer list
> 2. Operator can contact etcd; member list one too big for peers => create peer
> 3. Operator can contact etcd; members match peers; scale two big by one => add one member to etcd peer list
> 4. Operator can contact etcd; member list one too big for peers => create peer

Reply:
>  reconcile the EtcdPeer resources in Kubernetes to match the member list in etcd

you mean the peer status right? we're never going to be creating/updating `EtcdPeer` resource specs to match the membership API?

> If it cannot contact the etcd API, it will inspect the number of EtcdPeer resources in the cluster and if there fewer than the expected number from the EtcdCluster resource's spec.replicas value, then it will add new peer resources.

does it not also need to do this when the cluster is healthy?

Reply:
> […] we're never going to be creating/updating EtcdPeer resource specs to match the membership API?

We're going to do exactly that. During scale up the operator adds to the membership list in etcd, then on the next reconcile the operator sees that the membership list in etcd contains a member that it does not have a peer for, so it creates a new `EtcdPeer` to account for it. The etcd membership list is *the* canonical representation of the cluster. 

> does it not also need to do this when the cluster is healthy?

No. When the cluster is healthy we will scale up by adding a member to the etcd membership list *first*, and then react to that change to add the `EtcdPeer` resource. The behaviour for "unable to contact etcd" isn't really a scale-up at all. It's to bring a cluster in to quaracy so it can later interact with it properly.

Reply:
>    cluster controller: Create a new etcd peer
>    peer controller: Create a PVC & Pod for that peer
>      a) if the cluster is reachable - add it to member ship list
>      b) else - attempt bootstrap a new cluster with all existing peers

The problem comes that it's the 'wrong way around'. The etcd operations guide talks about the procedure for adding a new member [here](https://etcd.io/docs/v3.4.0/op-guide/runtime-configuration/#add-a-new-member).

> Adding a member is a two step process:
>  • Add the new member to the cluster via the HTTP members API, the gRPC members API, or the `etcdctl member add` command.
>  • Start the new member with the new cluster configuration, including a list of the updated members (existing members + the new member).

So in that example, you'd launch the peer *before* adding it to the membership API. We could try to solve this problem by using an init container of some kind (to check that the peer is in the cluster before launching it) but that would break in a bootstrap situation (or during a network disruption during a pod restart).

You could also think of having the peer controller decide, before launching the Pod, if it needs to add the member to the peer list. But that brings similar problems with the peer controller having to understand a lot about how to talk to the etcd *cluster*, needing specialist behaviour to detect a bootstrap situation (when should it launch anyway even if it can't launch, when should it wait?)

Reply:
>  I'd expect the custom resources to be the desired state of the world,

They _are_ the desired state of the world - it's just that the thing being modified is the membership API, and not the EtcdPeer resources (because as James notes, it's a two-step process to add a peer to the API where the *first* step is to register it with the membership API).

The overall architecture/design of the controller, and specifically the way that:

1) bootstrapping
2) scaling up
3) scaling down

is predicated on this assumption, and the reasoning for it was to closely mirror what etcd actually recommends.

By taking this approach, it actually makes the 'bootstrap' (aka the 'cold-start a cluster mode') far easier and safer to implement, as we can simply create EtcdPeers until the membership API is available, and then once the API _is_ available we can delete peers according to what the cluster actually expects to exist.

IMO, it's a matter of using the correct places as the source of truth. Replica count on the EtcdCluster resource should map to the membership API, whereas the EtcdPeer resources existence is dictated by the *contents* of the membership API (i.e. indirectly influenced by replica count). I think this is also safer, as it means we won't try and scale to situations that etcd's membership API would otherwise reject.

Reply:
I believe this discussion was resolved in video call. TL;DR the spec influences the membership list, the membership list influences the peers.